### PR TITLE
Gangams/logs 50k eps per node

### PR DESCRIPTION
**Pull request by ganga1980**
This PR has following changes 

1. AMA core agent integration on high log scale mode
2. Telemetry related to AMA core agent traces and high log scale mode
3.  Extend Config Map for high log scale mode
4.  Use high log scale config related to fluent-bit, such as threading, storage type
5. Extend liveness probe for AMA Core Agent when high log scale enabled
6. ARM template  updates related to high log scale

Onboarding steps
1.  Configure enabled = true under [agent_settings.high_log_scale] in ConfigMap
2.  Specify Microsoft-ContainerLogV2-HighScale  instead of Microsoft-ContainerLogV2 in ARM template parameter file 

Scale tests has been performed on VM SKU - Standard_D16s_v5

For 50k eps,  we would need minimum 2GB and 3Core.

Validated following scenarios with high log scale mode

1.  ContainerLogV2 with Microsoft-ContainerLogV2-HighScale stream on both Linux and Windows
2.  Here are memory usage numbers for  MDSD + AMA Core Agent on various log scales
             - No scale  MDSD (90MB) and AMACA (90MB)
             - 3K eps     MDSD(120MB) and AMACA(300MB)
             - 50K eps   MDSD(120MB) and AMACA(400MB)
3. HTTP proxy
4. AMPLS
5. Both Windows & Linux

Without high log scale mode (i.e. default ) 
 1. Validated on ARM64, Azure Linux and Windows nodes

Not supported high log scale scenarios - 
1. ARM64
2. HTTP proxy with CA cert

Resource Usage at 50K eps -  CPU is around 4 core and memory is 2.5GB, Disk IOPS - 570.

### CODE REVIEW

**Comment by ganga1980** [comment]
Max Memory and CPU Usage at 50K eps 

Max Memory Usage is 1.2GB and Max CPU Usage is 3.6 core
image

Logs/minute and logs/sec on scale cluster node

image

Logs Ingested vs Expected vs LogLossOrduplicates
image

image

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 1]
```diff
@@ -48,6 +59,15 @@ def substituteMultiline(multilineLogging, stacktraceLanguages, new_contents)
     return new_contents
 end
 
+def substituteStorageTotalLimitSize(new_contents)
+   if is_high_log_scale_mode?
+      new_contents = new_contents.gsub("${STORAGE_TOTAL_LIMIT_SIZE_MB}", "storage.total_limit_size        1000M")
```
Comment (wanlonghenry) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb]: do we want to change the new added config to this format in fluent-bit conf? #${STORAGE_TOTAL_LIMIT_SIZE_MB} storage.total_limit_size        1000M     and then here: new_contents = new_contents.gsub("#${STORAGE_TOTAL_LIMIT_SIZE_MB}", "")

File: build/linux/installer/conf/fluent-bit-common.conf [hunk 2]
```diff
@@ -77,13 +91,20 @@
...
+    Exclude log Information
+
 [OUTPUT]
     Name                            oms
     Alias                           oms_output
     EnableTelemetry                 true
     Retry_Limit                     10
     TelemetryPushIntervalSeconds    300
     Match                           oms.container.*
+    ${STORAGE_TOTAL_LIMIT_SIZE_MB}
```
Comment (wanlonghenry) [inline build/linux/installer/conf/fluent-bit-common.conf]: do we want to change the new added config to this format in fluent-bit conf? #${STORAGE_TOTAL_LIMIT_SIZE_MB} storage.total_limit_size 1000M

File: build/windows/installer/conf/fluent-bit-common.conf [hunk 3]
```diff
@@ -46,6 +46,7 @@
     Retry_Limit                     10
     TelemetryPushIntervalSeconds    300
     Match                           oms.container.*
+    ${STORAGE_TOTAL_LIMIT_SIZE_MB}
```
Comment (wanlonghenry) [inline build/windows/installer/conf/fluent-bit-common.conf]: do we want to change the new added config to this format in fluent-bit conf? #${STORAGE_TOTAL_LIMIT_SIZE_MB} storage.total_limit_size 1000M

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 4]
```diff
@@ -115,6 +145,33 @@ def substituteFluentBitPlaceHolders
       new_contents = new_contents.gsub("\n    ${TAIL_IGNORE_OLDER}\n", "\n")
     end
 
+    if is_high_log_scale_mode?
```
Comment (wanlonghenry) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb]: consider defining a function as the code logic is similar here: 
new_contents = update_config(new_contents, "TAIL_THREADED", "threaded on", enableFbitThreading)
new_contents = update_config(new_contents, "STORAGE_TYPE", "storage.type " + @default_high_log_scale_max_storage_type, storageType)
new_contents = update_config(new_contents, "MAX_STORAGE_CHUNKS_UP", "storage.max_chunks_up " + @default_high_log_scale_max_storage_chunks_up, storageMaxChunksUp)

File: build/linux/installer/conf/fluent-bit.conf [hunk 5]
```diff
@@ -3,6 +3,7 @@
     ${SERVICE_FLUSH_INTERVAL}
     HTTP_Server   Off
     Daemon        Off
+    ${MAX_STORAGE_CHUNKS_UP}
```
Comment (wanlonghenry) [inline build/linux/installer/conf/fluent-bit.conf:6]: similar to above comment

File: build/linux/installer/conf/fluent-bit.conf [hunk 6]
```diff
@@ -24,6 +25,7 @@
     DB /var/log/omsagent-fblogs.db
     DB.Sync Off
     Parser docker
+    ${STORAGE_TYPE}
```
Comment (wanlonghenry) [inline build/linux/installer/conf/fluent-bit.conf:28]: similar to above comment

File: source/plugins/go/src/extension/extension.go [hunk 7]
```diff
@@ -153,13 +171,27 @@ func getDataTypeToStreamIdMapping(hasNamedPipe bool) (map[string]string, error)
...
+			// } else {
+				if hasNamedPipe {
+					datatypeOutputStreamMap[dataType] = outputStreamDefinitions[outputStreamID.(string)].NamedPipe
+				} else {
+					datatypeOutputStreamMap[dataType] = outputStreamID.(string)
+				}
+		//    }
 		}
 	}
+
```
Comment (wanlonghenry) [inline source/plugins/go/src/extension/extension.go]: remove empty line

File: source/plugins/go/src/extension/extension.go [hunk 8]
```diff
@@ -153,13 +171,27 @@ func getDataTypeToStreamIdMapping(hasNamedPipe bool) (map[string]string, error)
 	for _, extensionConfig := range extensionConfigs {
 		outputStreams := extensionConfig.OutputStreams
 		for dataType, outputStreamID := range outputStreams {
-			if hasNamedPipe {
-				datatypeOutputStreamMap[dataType] = outputStreamDefinitions[outputStreamID.(string)].NamedPipe
-			} else {
-				datatypeOutputStreamMap[dataType] = outputStreamID.(string)
-			}
+			////BUG - if DCR has both streams Microsoft-ContainerLogV2 & Microsoft-ContainerLogV2-HighScale, AMA ingests duplicate records hence ensure only stream in the DCR dependening on high log scale or not
+			// if isHighLogScaleMode() && dataType == containerLogV2DataType {
```
Comment (wanlonghenry) [inline source/plugins/go/src/extension/extension.go]: this looks like a TODO item, could you explain more on this part of logic as it does not change existing code

**Comment by ganga1980** [comment]
Disk IO charts at 50K logs/sec/node.

Total Max Read & Writes/sec is 570 at peak. At peak, Read/sec is 329 and writes/sec is 242.

image

**Comment by ganga1980** [comment]
latest test run with MDSD & AMACA RC image - 
image
image

**Comment by ganga1980** [comment]
On default scenario (without high log scale mode) on MSI mode on x64, ARM64 and windows node  -  

Daemonset Memory Usage (with & without AMA version update)
image

Daemonset CPU Usage (with & without AMA version update)
image

Replicaset CPU Usage
image

Replicaset Memory Usage
image

On default scenario (without high log scale mode) on Legacy mode on x64, ARM64 and windows node  -  

Daemonset Memory Usage (with & without AMA version update)
image

Deamonset CPU Usage  (with & without AMA version update)

image

Replicaset CPU Usage (with & without AMA version update)
image

Replicaset memory usage(with & without AMA version update)

image

File: build/linux/installer/conf/fluent-bit-common.conf [hunk 9]
```diff
@@ -77,10 +91,25 @@
...
     EnableTelemetry                 true
     Retry_Limit                     10
     TelemetryPushIntervalSeconds    300
     Match                           oms.container.*
+    #${OMS_STORAGE_TOTAL_LIMIT_SIZE_MB}
+
+[OUTPUT]
+    Name                            oms
+    Alias                           oms_input_plugin_output
+    EnableTelemetry                 true
```
Comment (pfrcks) [inline build/linux/installer/conf/fluent-bit-common.conf]: this section looks to be from the old version of input plugin. can you remove it? strange rebase did not catch it

File: build/linux/installer/scripts/dcr-config-parser.rb [hunk 10]
```diff
@@ -1,17 +1,17 @@
 #!/usr/local/bin/ruby
```
Comment (pfrcks) [inline build/linux/installer/scripts/dcr-config-parser.rb:1]: why are changes to file part of the PR?

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 11]
```diff
@@ -48,6 +60,15 @@ def substituteMultiline(multilineLogging, stacktraceLanguages, new_contents)
     return new_contents
 end
 
+def substituteStorageTotalLimitSize(new_contents)
+   if is_high_log_scale_mode?
+      new_contents = new_contents.gsub("#${OMS_STORAGE_TOTAL_LIMIT_SIZE_MB}", "storage.total_limit_size  " + @default_high_log_scale_max_storage_total_limit_size)
+   else
+      new_contents = new_contents.gsub("\n    #${OMS_STORAGE_TOTAL_LIMIT_SIZE_MB}\n", "\n")
```
Comment (pfrcks) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb]: any reason for using the OMS_ naming pattern instead of AZMON?

File: build/linux/installer/conf/fluent-bit-common.conf (hunk 9 above)
Comment (ganga1980) [reply build/linux/installer/conf/fluent-bit-common.conf]: > this section looks to be from the old version of input plugin. can you remove it? strange rebase did not catch it

Fixed. can you please double check if there are any other issues.

File: build/linux/installer/scripts/dcr-config-parser.rb (hunk 10 above)
Comment (ganga1980) [reply build/linux/installer/scripts/dcr-config-parser.rb:1]: > why are changes to file part of the PR?

> there is code comment  on this change. "# If the DCR has GIG and ODS streams,then id will be either in the format ContainerInsightsExtension:ods-* or ContainerInsightsExtension:gigl-dce-*"

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 12]
```diff
@@ -79,8 +134,18 @@ def substituteFluentBitPlaceHolders
     windowsFluentBitDisabled = ENV["AZMON_WINDOWS_FLUENT_BIT_DISABLED"]
     kubernetesMetadataCollection = ENV["AZMON_KUBERNETES_METADATA_ENABLED"]
     annotationBasedLogFiltering = ENV["AZMON_ANNOTATION_BASED_LOG_FILTERING"]
+    storageMaxChunksUp = ENV["FBIT_STORAGE_MAX_CHUNKS_UP"]
```
Comment (jatakiajanvi12) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb:137]: What happens when all this is empty?

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 13]
```diff
@@ -79,8 +134,18 @@ def substituteFluentBitPlaceHolders
...
+    storageType = ENV["FBIT_STORAGE_TYPE"]
```
Comment (jatakiajanvi12) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb:138]: should we have any validation for storage type? Because that will be something specific right for fluent-bit?

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 14]
```diff
@@ -63,6 +84,40 @@ def substituteResourceOptimization(resourceOptimizationEnabled, new_contents)
...
+        puts "Since high log scale mode configured hence using storage.type: #{@default_high_log_scale_max_storage_type} for tail plugin"
+      elsif !storageType.nil? && !storageType.empty?
+        new_contents = new_contents.gsub("#${AZMON_STORAGE_TYPE}", "storage.type " + storageType)
+      else
+        new_contents = new_contents.gsub("\n    #${AZMON_STORAGE_TYPE}\n", "\n")
+      end
+
+      if is_high_log_scale_mode?
+        new_contents = new_contents.gsub("#${AZMON_MAX_STORAGE_CHUNKS_UP}", "storage.max_chunks_up " + @default_high_log_scale_max_storage_chunks_up)
+        puts "Since high log scale mode configured hence using storage.max_chunks_up: #{@default_high_log_scale_max_storage_chunks_up} for tail plugin"
```
Comment (jatakiajanvi12) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb]: Should we make these log statements smaller? As they will all be in our telemetry as well

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb [hunk 15]
```diff
@@ -115,6 +180,8 @@ def substituteFluentBitPlaceHolders
       new_contents = new_contents.gsub("\n    ${TAIL_IGNORE_OLDER}\n", "\n")
     end
 
+    new_contents = substituteHighLogScaleConfig(enableFbitThreading, storageType,  storageMaxChunksUp, new_contents)
```
Comment (jatakiajanvi12) [inline build/common/installer/scripts/fluent-bit-conf-customizer.rb:183]: should we be substituting this for geneva as well?

File: source/plugins/go/src/extension/extension.go [hunk 16]
```diff
@@ -153,11 +153,11 @@ func getDataTypeToStreamIdMapping(hasNamedPipe bool) (map[string]string, error)
 	for _, extensionConfig := range extensionConfigs {
 		outputStreams := extensionConfig.OutputStreams
 		for dataType, outputStreamID := range outputStreams {
-			if hasNamedPipe {
-				datatypeOutputStreamMap[dataType] = outputStreamDefinitions[outputStreamID.(string)].NamedPipe
-			} else {
-				datatypeOutputStreamMap[dataType] = outputStreamID.(string)
-			}
+				if hasNamedPipe {
+					datatypeOutputStreamMap[dataType] = outputStreamDefinitions[outputStreamID.(string)].NamedPipe
```
Comment (jatakiajanvi12) [inline source/plugins/go/src/extension/extension.go:157]: why is this change necessary?

File: kubernetes/linux/main.sh [hunk 17]
```diff
@@ -5,6 +5,46 @@ startTime=$(date +%s)
...
+      } >> ~/.bashrc
+
+      source ~/.bashrc
+      /opt/microsoft/azure-mdsd/bin/amacoreagent -c $AMACAConfigFilePath --configport $PA_CONFIG_PORT --amacalog $AMACALogFilePath > /dev/null 2>&1 &
+
+      waitforlisteneronTCPport "$PA_FLUENT_SOCKET_PORT" "$WAITTIME_PORT_13000"
+      waitforlisteneronTCPport "$PA_CONFIG_PORT" "$WAITTIME_PORT_12563"
+      # Extract AMACoreAgent version from log file
+      version=""
+      if [ -d "$AMACALogFileDir" ]; then
```
Comment (jatakiajanvi12) [inline kubernetes/linux/main.sh:39]: should we instead add this in setup.sh? similar to what we are doing for other packages?

Comment (ganga1980) [reply kubernetes/linux/main.sh:39]: > should we instead add this in setup.sh? similar to what we are doing for other packages?

> since  AMACA binary part of the same  azure-mdsd package, and we cant get the version of the AMACA and AMACA binary doesnt support --version command properly so we can only get from logs.

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb (hunk 12 above)
Comment (ganga1980) [reply build/common/installer/scripts/fluent-bit-conf-customizer.rb:137]: > What happens when all this is empty?

> if this empty, there is no change. we only honor if its not nil & not empty and   greater than 0.

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb (hunk 13 above)
Comment (ganga1980) [reply build/common/installer/scripts/fluent-bit-conf-customizer.rb:138]: > should we have any validation for storage type? Because that will be something specific right for fluent-bit?

Yes, for all the config parameters and there is an validation . For example, see  - https://github.com/microsoft/Docker-Provider/blob/062c9b4ec9286702e833f6bb39444187f04eacdb/build/common/installer/scripts/tomlparser-agent-config.rb#L252

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb (hunk 15 above)
Comment (ganga1980) [reply build/common/installer/scripts/fluent-bit-conf-customizer.rb:183]: for geneva path handled in geneva-conf-customizer and through ${TAIL_THREADED} which is different from AZMON path.

File: source/plugins/go/src/extension/extension.go (hunk 16 above)
Comment (ganga1980) [reply source/plugins/go/src/extension/extension.go:157]: > why is this change necessary?

There is no code  change here. its just white space change.

File: build/common/installer/scripts/fluent-bit-conf-customizer.rb (hunk 14 above)
Comment (ganga1980) [reply build/common/installer/scripts/fluent-bit-conf-customizer.rb]: > Should we make these log statements smaller? As they will all be in our telemetry as well

> as we discussed, we would need to optimize our agent logs to reduce the telemetry data size. just changing wouldnt make significant but anyway optimized it.

File: kubernetes/linux/main.sh [hunk 18]
```diff
@@ -5,6 +5,46 @@ startTime=$(date +%s)
...
+         echo "export CounterDataReportFrequencyInMinutes=$CounterDataReportFrequencyInMinutes"
+      } >> ~/.bashrc
+
+      source ~/.bashrc
+      /opt/microsoft/azure-mdsd/bin/amacoreagent -c $AMACAConfigFilePath --configport $PA_CONFIG_PORT --amacalog $AMACALogFilePath > /dev/null 2>&1 &
+
+      waitforlisteneronTCPport "$PA_FLUENT_SOCKET_PORT" "$WAITTIME_PORT_13000"
+      waitforlisteneronTCPport "$PA_CONFIG_PORT" "$WAITTIME_PORT_12563"
+      # Extract AMACoreAgent version from log file
+      version=""
```
Comment (pfrcks) [inline kubernetes/linux/main.sh:38]: can't we do something like amacoreagent -v?

Comment (ganga1980) [reply kubernetes/linux/main.sh:38]: > can't we do something like amacoreagent -v?

No, that doesn't work. I already tried.

**Review by pfrcks — APPROVED** [review]
